framework/resource/internal/wrapper.go

```diff
@@ -1,7 +1,29 @@
 package internal
 
-import "github.com/giantswarm/operatorkit/framework"
+import (
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/framework"
+)
 
 type Wrapper interface {
-	Underlying() framework.Resource
+	Wrapped() framework.Resource
+}
+
+func Underlying(r framework.Resource) (framework.Resource, error) {
+	i := 0
+
+	for {
+		wrapper, ok := r.(Wrapper)
+		if ok {
+			r = wrapper.Wrapped()
+		} else {
+			return r, nil
+		}
+
+		// When more that 1000 interations, assume infifinite loop.
+		i++
+		if i > 1000 {
+			return nil, microerror.Maskf(loopDetectedError, "unwrapped 1000 times, assuming infite loop: resource = %s", r.Name())
+		}
+	}
```

Comment:
Why does this have to be so complicated? What is wrong with the current implementation?

Reply:
The current (simplified) implementation doesn't unwrap nested resources. I wouldn't call it complex. It just unwraps a resource wrapped by possibly many wrappers. Which in our case is retryresource and metricsresource (many :D), but it has to be ready for possible more wrappers.

When we have CRUDResourceOps and CRUDResource it is difficult to set Underling in the right place.

It is also difficult to explain to users why you need to add Underlying method. And why it should return itself. It took me time to realize that at first place. We're doing the workshop, going to write a blogpost. We have 65 stars (which is a lot in operatokits word) - I'd like to make the thing more user friendly, easier to digest and spread the word.